pkg/messaging/mqtt/setup_test.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package mqtt_test
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+	"testing"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/gogo/protobuf/proto"
+	logg "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	mqtt_pubsub "github.com/mainflux/mainflux/pkg/messaging/mqtt"
+	"github.com/ory/dockertest/v3"
+)
+
+var (
+	publisher messaging.Publisher
+	pubsub    messaging.PubSub
+
+	logger logg.Logger
+
+	address = "tcp://mqtt.eclipseprojects.io:1883"
+)
+
+const (
+	username = "mainflux-mqtt"
+	qos      = 2
+)
+
+func TestMain(m *testing.M) {
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	container, err := pool.Run("eclipse-mosquitto", "1.6.13", nil)
+	if err != nil {
+		log.Fatalf("Could not start container: %s", err)
+	}
+	handleInterrupt(m, pool, container)
+
+	address = fmt.Sprintf("%s:%s", "localhost", container.GetPort("1883/tcp"))
+	pool.MaxWait = 120 * time.Second
+	if err := pool.Retry(func() error {
+		publisher, err = mqtt_pubsub.NewPublisher(address, 30*time.Second)
+		return err
+	}); err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	logger, err = logg.New(os.Stdout, "error")
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	if err := pool.Retry(func() error {
+		pubsub, err = mqtt_pubsub.NewPubSub(address, "", 30*time.Second, logger)
+		return err
+	}); err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	code := m.Run()
+	if err := pool.Purge(container); err != nil {
+		log.Fatalf("Could not purge container: %s", err)
+	}
+
+	os.Exit(code)
+
+	defer func() {
+		err = publisher.Close()
+		if err != nil {
+			log.Fatalf(err.Error())
+		}
+		err = pubsub.Close()
+		if err != nil {
+			log.Fatalf(err.Error())
+		}
+	}()
+}
+
+func handleInterrupt(m *testing.M, pool *dockertest.Pool, container *dockertest.Resource) {
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		if err := pool.Purge(container); err != nil {
+			log.Fatalf("Could not purge container: %s", err)
+		}
+		os.Exit(0)
+	}()
+}
+
+func mqttHandler(h messaging.MessageHandler) mqtt.MessageHandler {
+	return func(c mqtt.Client, m mqtt.Message) {
+		var msg messaging.Message
+		if err := proto.Unmarshal(m.Payload(), &msg); err != nil {
+			logger.Warn(fmt.Sprintf("Failed to unmarshal received message: %s", err))
+			msg.Payload = m.Payload()
+			if err := h.Handle(msg); err != nil {
+				logger.Warn(fmt.Sprintf("Failed to handle Mainflux message: %s", err))
+			}
+			return
+		}
+		if err := h.Handle(msg); err != nil {
+			logger.Warn(fmt.Sprintf("Failed to handle Mainflux message: %s", err))
+		}
+	}
+}
+
+func newClient(address, id string, timeout time.Duration) (mqtt.Client, error) {
```

Comment:
Why are you using using here?
```go
func newClient(address, id string, timeout time.Duration) (mqtt.Client, error)
```
I thought when you instantiate publisher or pubsub it will automatically create a client and manage it for you.

Reply:
I am also testing publisher and subscribe-unsubscribe separately.
So, to test only `publish`, I use default mqtt.client.Subscribe(), instead of using the pubsub implementation.
Similarly for testing only subscribe, that's why I've created this newClient function. Same goes for mqttHandler in setup_test

Reply:
@AryanGodara Please do not mark it as resolved yet but wait for @0x6f736f646f to address the answer and let him close the discussion if the answer is OK.